### PR TITLE
Enforce per-query chunks limit earlier when streaming chunks from ingesters to queriers

### PR DESCRIPTION
**Pull request by charleskorn**
#### What this PR does

This PR enforces the per-query chunks limit earlier during query evaluation when streaming chunks from ingesters to queriers. This helps prevent excessive memory utilisation or OOMing while processing a query that will exceed the per-query chunks limit.

~~Before this can be merged, https://github.com/grafana/mimir-prometheus/pull/513 needs to be merged.~~ ✅ Merged

#### Which issue(s) this PR fixes or relates to

(none)

#### Checklist

- [x] Tests updated
- [n/a] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/ingester/client/streaming.go [hunk 1]
```diff
@@ -132,11 +132,7 @@ func (s *SeriesChunksStreamReader) StartBuffering() {
 				}
 			}
 
-			if err := s.queryLimiter.AddChunks(chunkCount); err != nil {
-				onError(err)
-				return
-			}
-
+			// The chunk count limit is enforced earlier, while we're reading series labels, so we don't need to do that here.
```
Comment (dimitarvdimitrov) [inline pkg/ingester/client/streaming.go:135]: It seems like you can remove the intermediate `chunkCount` and always increment `totalChunks`

**Review by dimitarvdimitrov — APPROVED** [review]
LGTM

**Review by pracucci — APPROVED** [review]
LGTM.

@dimitarvdimitrov I agree with your comment, but Charles asked to merge this PR this week so that it will be included in the next week release. Since the comment is non-blocking I'm going to merge this PR and then @charleskorn can address it in a follow up PR.

File: pkg/ingester/client/streaming.go (hunk 1 above)
Comment (charleskorn) [reply pkg/ingester/client/streaming.go:135]: Good point, addressed in https://github.com/grafana/mimir/pull/5447

**Pull request merged**